src/Management/test/Endpoint.Test/SpringBootAdminClient/AppUrlCalculatorTest.cs

```diff
@@ -0,0 +1,383 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.AspNetCore.Hosting.Server;
+using Microsoft.AspNetCore.Hosting.Server.Features;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Steeltoe.Common.Net;
+using Steeltoe.Management.Endpoint.Configuration;
+using Steeltoe.Management.Endpoint.SpringBootAdminClient;
+
+namespace Steeltoe.Management.Endpoint.Test.SpringBootAdminClient;
+
+public sealed class AppUrlCalculatorTest
+{
+    private const string ListenNonSecurePort1 = "4444";
+    private const string ListenNonSecurePort2 = "5555";
+    private const string ListenSecurePort1 = "6666";
+    private const string ListenSecurePort2 = "7777";
+    private const string ManagementPort = "8888";
+    private const string OverriddenPort = "9999";
+
+    [Fact]
+    public void Selects_default_binding_when_nothing_configured()
+    {
+        IConfiguration configuration = new ConfigurationBuilder().Build();
+        using ServiceProvider serviceProvider = BuildServiceProvider(configuration);
+
+        var calculator = serviceProvider.GetRequiredService<AppUrlCalculator>();
+        var options = serviceProvider.GetRequiredService<IOptions<SpringBootAdminClientOptions>>();
+        string? url = calculator.AutoDetectAppUrl(options.Value);
+
+        url.Should().Be($"http://{FakeDomainNameResolver.HostName}:5000");
+    }
+
+    [Fact]
+    public void Prefers_https_binding_when_multiple_urls_configured()
+    {
+        var appSettings = new Dictionary<string, string?>
+        {
+            ["urls"] = $"http://dontcare:{ListenNonSecurePort1};https://dontcare:{ListenSecurePort1};https://dontcare:{ListenSecurePort2}"
+        };
+
+        IConfiguration configuration = new ConfigurationBuilder().AddInMemoryCollection(appSettings).Build();
+        using ServiceProvider serviceProvider = BuildServiceProvider(configuration);
+
+        var calculator = serviceProvider.GetRequiredService<AppUrlCalculator>();
+        var options = serviceProvider.GetRequiredService<IOptions<SpringBootAdminClientOptions>>();
+        string? url = calculator.AutoDetectAppUrl(options.Value);
+
+        url.Should().Be($"https://{FakeDomainNameResolver.HostName}:{ListenSecurePort1}");
```

Comment:
Maybe I'm missing something, but I would think that since the value in `urls` is not a wildcard the hostname should be "dontcare"... Won't this cause additional certificate issues? -- We know SBA in a container already didn't trust the ASP.NET Core cert, and can't reach `localhost` anyway, so maybe that's not a great example, but now the host also won't match

Reply:
This test case is intentional, to express any hostname is meaningless. See the docs at https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-9.0#url-formats:

> Anything not recognized as a valid IP address or localhost is treated as a wildcard that binds to all IPv4 and IPv6 addresses. Some people like to use * or + to be more explicit.

The hostname has no meaning in ASP.NET, so it would be wrong if it meant something in Steeltoe. Users wouldn't expect that.

Reply:
Huh, based on the different docs it looks like that changed in 8.0, I either didn't know that or forgot about it...
In any case, it looks like localhost still gets special treatment in ASP.NET Core, but based on the tests we would register the actual hostname with SBA, that seems like it shouldn't work, but does it?

Reply:
ASP.NET has always worked like that, as far as I'm aware. The [documentation for .NET 5](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-5.0#url-formats) uses slightly different wording:
> Host names, *, and +, aren't special. Anything not recognized as a valid IP address or localhost binds to all IPv4 and IPv6 IPs.

I'm going to change the logic so that `localhost` and IP addresses from bindings are preferred over resolving the local host/IP.